### ISSUE TITLE
Optimize whichmut loss computation with vectorized implementation

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/netam#158]
## Summary
The current `compute_whichmut_loss_batch()` function in `netam/whichmut_trainer.py` uses explicit loops for clarity and correctness verification. A vectorized implementation using PyTorch advanced indexing and masking operations would provide significant performance improvements for large datasets.

## Current Implementation
The current loop-based approach (lines 223-257 in `whichmut_trainer.py`):
```python
log_probs = []
for seq_idx in range(N):
    for codon_pos in range(L_codon):
        if codon_mutation_indicators[seq_idx, codon_pos] and masks[seq_idx, codon_pos]:
            # Extract mutation details
            parent_codon_idx = codon_parents_idxss[seq_idx, codon_pos]
            child_codon_idx = codon_children_idxss[seq_idx, codon_pos]
            
            # Compute probability and accumulate
            # ... (detailed computation)
            log_probs.append(torch.log(prob + 1e-10))
```

**Performance characteristics:**
- ✅ **Readable**: Easy to understand and debug
- ✅ **Correct**: Mathematically verified with reference implementation  
- ❌ **Slow**: O(N × L_codon) with Python loops
- ❌ **Memory inefficient**: Creates intermediate lists

## Proposed Vectorized Implementation
Based on the TODO comment from issue #152, implement the following vectorized approach:

```python
def compute_whichmut_loss_batch_vectorized(
    selection_factors: torch.Tensor,
    neutral_rates_tensor: torch.Tensor,
    codon_parents_idxss: torch.Tensor,
    codon_children_idxss: torch.Tensor,
    codon_mutation_indicators: torch.Tensor,
    aa_parents_idxss: torch.Tensor,
    masks: torch.Tensor,
) -> torch.Tensor:
    """Vectorized version of whichmut loss computation for performance."""
    
    # Convert selection factors from log space to linear space
    linear_selection_factors = torch.exp(selection_factors)
    
    # Compute normalization constants Z_j for each codon position
    normalization_constants = compute_normalization_constants(
        linear_selection_factors, neutral_rates_tensor, aa_parents_idxss
    )
    
    # 1. Get indices of mutated positions using advanced indexing
    mutated_mask = codon_mutation_indicators & masks  # (N, L_codon)
    mutated_indices = torch.where(mutated_mask)  # (seq_indices, codon_indices)
    
    # Early return if no mutations
    if len(mutated_indices[0]) == 0:
        return torch.tensor(0.0, device=selection_factors.device, requires_grad=True)
    
    # 2. Extract parent/child codon indices for mutated positions only
    seq_idx, codon_idx = mutated_indices
    mut_parent_codons = codon_parents_idxss[mutated_indices]  # (n_mutations,)
    mut_child_codons = codon_children_idxss[mutated_indices]   # (n_mutations,)
    
    # 3. Advanced indexing to get neutral rates for observed mutations
    neutral_rates = neutral_rates_tensor[seq_idx, codon_idx, mut_parent_codons, mut_child_codons]
    
    # 4. Map child codons to AA indices and extract selection factors
    child_aa_indices = aa_idx_of_flat_codon_idx_vectorized(mut_child_codons)  # Needs vectorized version
    selection_factors_masked = linear_selection_factors[seq_idx, codon_idx, child_aa_indices]
    
    # 5. Get normalization constants for mutated positions
    Z_j_masked = normalization_constants[mutated_indices]
    
    # 6. Compute probabilities and log likelihood vectorized
    probs = (neutral_rates * selection_factors_masked) / Z_j_masked
    log_likelihood = torch.log(probs + 1e-10).sum()
    
    return -log_likelihood
```

### 1. Vectorized Codon-to-AA Mapping
The current `aa_idx_of_flat_codon_idx()` function processes single codon indices. Need a vectorized version:

```python
def aa_idx_of_flat_codon_idx_vectorized(codon_idxs: torch.Tensor) -> torch.Tensor:
    """Vectorized version of codon to AA index mapping.
    
    Args:
        codon_idxs: (n_mutations,) tensor of codon indices
        
    Returns:
        (n_mutations,) tensor of corresponding AA indices
    """
    # Handle ambiguous codons (>= 64)
    ambiguous_mask = codon_idxs >= 64
    
    # For valid codons, use the indicator matrix
    valid_codon_idxs = torch.clamp(codon_idxs, 0, 63)
    codon_rows = CODON_AA_INDICATOR_MATRIX[valid_codon_idxs]  # (n_mutations, 20)
    
    # Check for stop codons (all zeros in the matrix)
    stop_mask = codon_rows.sum(dim=1) == 0
    
    # Get AA indices using argmax
    aa_idxs = codon_rows.argmax(dim=1)
    
    # Set ambiguous and stop codons to index 20
    aa_idxs[ambiguous_mask | stop_mask] = 20
    
    return aa_idxs
```

### 2. Vectorized Normalization Constants
The current `compute_normalization_constants()` also uses loops and could benefit from vectorization, though this is more complex due to the double loop over codon indices.

## Performance Benefits
**Expected improvements:**
- **Speed**: 10-100x faster for large datasets (eliminates Python loops)
- **Memory**: More efficient tensor operations, better GPU utilization
- **Scalability**: Linear scaling with number of actual mutations rather than sequence length

**Benchmark scenarios:**
- Small datasets (N=100, L_codon=50): Moderate improvement
- Large datasets (N=10,000, L_codon=500): Significant improvement  
- Sparse mutations: Major improvement (only processes actual mutations)

## Implementation Plan
1. **Phase 1**: Implement `aa_idx_of_flat_codon_idx_vectorized()`
2. **Phase 2**: Implement vectorized loss computation
3. **Phase 3**: Add comprehensive tests comparing loop vs vectorized results
4. **Phase 4**: Benchmark performance across different dataset sizes
5. **Phase 5**: Consider vectorizing normalization constants computation

## Testing Requirements
- ✅ **Mathematical equivalence**: Vectorized version must produce identical results to loop version
- ✅ **Edge cases**: Handle no mutations, all mutations, boundary conditions  
- ✅ **Performance benchmarks**: Measure speedup across dataset sizes
- ✅ **Memory profiling**: Ensure memory usage is reasonable

## Technical Challenges
1. **Complex indexing**: Advanced indexing with multiple dimensions requires careful tensor shape management
2. **Codon mapping**: Vectorizing genetic code lookups while handling stop codons and ambiguous codons
3. **Normalization constants**: The Z_j computation involves nested loops that are harder to vectorize
4. **Memory vs speed**: Vectorization may use more memory for intermediate tensors

## References
- Original issue #152: Whichmut Trainer and Loss Function implementation
- Current implementation: `netam/whichmut_trainer.py:compute_whichmut_loss_batch()`
- PyTorch advanced indexing: https://pytorch.org/docs/stable/advanced_indexing.html

This optimization would significantly improve training speed for large-scale whichmut model training while maintaining mathematical correctness.
